server/controllers/apiController.js

```diff
@@ -4,60 +4,76 @@
-              var review = reviews[j];
-              if (review.text.match(regex)) { //TODO: improve regex matching
-                filteredBody.places.push({
-                  name: placeDetails.name,
-                  address: placeDetails['formatted_address']
-                });
-                break;
-              }
-            }
-            counter++;
-            if (counter === places.length) {
-              res.json(filteredBody); //also needs to pass page token back to client for further requests
-            } //TODO: move it inside callback above
-          })
-          .on('error', function(error) {
-            //TODO: handle error
-            counter++;
-            if (counter === places.length) {
-              res.json(filteredBody); //also needs to pass page token back to client for further requests
-            } 
-          })
+      response.on('data', function(chunk) { //layer 2 on 'data'
+        body.push(chunk);
+      }).on('end', function() { //layer 2 on 'end'
+        body = JSON.parse(Buffer.concat(body).toString());
+        var filteredBody = {};
+        filteredBody.places = [];
+        var places = body.results;
+        var counter = 0; //ensure server only sends back filteredBody if all places have been processed
+        for (var i = 0; i < places.length; i++) {
+          var place = places[i];
+          var placeid = place['place_id'];
 
-    })
-    .on('error', function(error) {
+          request.get('https://maps.googleapis.com/maps/api/place/details/json?' + 'key=' + GOOGLE_PLACES_API_KEY + '&placeid=' + placeid)
+            .on('response', function(response) { //layer 3 on 'response'
+              var body = [];
+              response.on('data', function(chunk) { //layer 4 on 'data'
+                body.push(chunk);
+              }).on('end', function() { //layer 4 on 'end'
+                body = JSON.parse(Buffer.concat(body).toString());
+                var placeDetails = body.result;
+                var reviews = placeDetails.reviews;
+                if (reviews) {
+                  for (var j = 0; j < reviews.length; j++) {
+                    var review = reviews[j];
+                    if (review.text.match(regex)) { //TODO: improve regex matching
+                      filteredBody.places.push({
+                        name: placeDetails.name,
+                        address: placeDetails['formatted_address']
+                      });
+                      break;
+                    }
+                  }
+                }
+                counter++;
+                if (counter === places.length) {
+                  res.json(filteredBody); //also needs to pass page token back to client for further requests
+                }
+              }); //end of layer 4 on 'end'
+            }) //end of layer 3 on 'response'
+            .on('error', function(error) { //layer 3 on 'error'
+              //TODO: handle error
+              counter++;
```

Comment:
@lauracurley mentioned that we could use `Promise.All` (or something like this) instead of the counter.